### ISSUE TITLE
Client stops consuming certain partitions

### ISSUE THREAD

**Issue by stampy88** [twmb/franz-go#507]
Hi, 

I am using the Franz library to consume messages from an Azure Event Hub topic. I have 2 consumers per region, and the Event Hub in the central region. My consumer is based on the [[goroutine_per_partition_consuming](https://github.com/twmb/franz-go/tree/master/examples/goroutine_per_partition_consuming)](https://github.com/twmb/franz-go/blob/master/examples/goroutine_per_partition_consuming/manual_commit/main.go) with manual committing, but I am not blocking a rebalance on polls. I am also using partition "flow control" to stop fetching partitions that have a backlog of records to process. I do this to ensure the polling go routine does not block waiting to send records to a worker. 

What I am seeing is that sometimes the fetch no longer returns records for a partition, but I haven't been able to pinpoint why. I am wondering if it is a race condition between the `pause/resume` and rebalancing. I am hoping by sharing my code, you'll be able to pinpoint whether I am doing something wrong, or whether there is a race condition in the library that is causing this odd behavior. Please note that there are some interfaces introduced in the code below for unit testing purposes.

```
package franz

import (
	"context"
	"sync"

	"github.com/XXX/log"
	"github.com/twmb/franz-go/pkg/kgo"
)

// newCoordinator will create a new coordinator that will deliver events to the specified [RecordListener]
func newCoordinator(logger log.Logger, l RecordListener) *coordinator {
	const defaultBufferSize = 10
	return newCoordinatorWithBufferSize(logger, l, defaultBufferSize)
}

func newCoordinatorWithBufferSize(logger log.Logger, l RecordListener, size int) *coordinator {
	return &coordinator{
		logger:   logger,
		listener: l,

		mu:         &sync.RWMutex{},
		workers:    make(map[TopicPartition]*worker),
		bufferSize: size,
		// we will pause fetching for a partition if the channel for a worker is at least 50% capacity
		pauseThreshold: 0.50 * float32(size),
	}
}

// a coordinator is used for managing a Kafka consumer group member and sending records to workers it is managing
type coordinator struct {
	logger log.Logger

	// a listener is the application code that a worker will deliver a batch of records to
	listener RecordListener

	// mu is used to guard the workers map since it is accessed in the run loop and via the consumer group management
	// callbacks
	mu      *sync.RWMutex
	workers map[TopicPartition]*worker

	// bufferSize is used to size the channel that is used to send records to a worker
	bufferSize int
	// pauseThreshold is the number of items in the records channel over which we will pause fetching data for this
	// [TopicPartition]/worker
	pauseThreshold float32
}

// onPartitionsAssigned is called when a group is joined after partitions are assigned before fetches for those
// partitions begin. we use it to create workers based upon the topic/partitions we have been assigned.
// NOTE - This function is not called concurrent with revoke or lost. This function can be called at any time
// while we are polling or processing records.
func (c *coordinator) onPartitionsAssigned(ctx context.Context, client *kgo.Client, assigned map[string][]int32) {
	if len(assigned) == 0 {
		return
	}
	c.handlePartitionsAssigned(ctx, client, assigned)
}

func (c *coordinator) handlePartitionsAssigned(ctx context.Context, client *kgo.Client, assigned map[string][]int32) {
	// this wait group is used to ensure the workers are running before returning
	var wg sync.WaitGroup
	defer wg.Wait()

	for topic, partitions := range assigned {
		for _, partition := range partitions {
			assignment := TopicPartition{topic, partition}

			w := c.newWorker(c.logger, client, assignment)

			c.mu.Lock()
			c.workers[assignment] = w
			c.mu.Unlock()
			wg.Add(1)

			go func() {
				wg.Done()
				w.run(ctx)
			}()
		}
	}
}

func (c *coordinator) newWorker(clog log.Logger, client *kgo.Client, assignment TopicPartition) *worker {
	return &worker{
		logger:     clog,
		client:     client,
		listener:   c.listener,
		assignment: assignment,

		stop:    make(chan struct{}),
		stopped: make(chan struct{}),
		// we make it a buffered channel so that the [coordinator.doRun] loop is not blocked while waiting
		// for a worker to pick up the records
		records: make(chan []*kgo.Record, c.bufferSize),

		// we want to resume fetching for this partition if the *used* channel capacity is <= 25%
		resumeThreshold: 0.25 * float32(c.bufferSize),
	}
}

// onPartitionsLost is called once this group member has partitions lost or revoked. We use it to stop any workers
// whose partitions we are losing. Since the worker is doing the committing, the coordinator will send a stop signal
// to each worker and wait until each is stopped before preceding. This blocking action is required so that the
// rebalance doesn't proceed until we have committed our work.
func (c *coordinator) onPartitionsLost(ctx context.Context, _ *kgo.Client, lost map[string][]int32) {
	if len(lost) == 0 {
		return
	}
	c.handlePartitionsLost(lost)
}

func (c *coordinator) handlePartitionsLost(lost map[string][]int32) {
	var wg sync.WaitGroup
	defer wg.Wait()

	for topic, partitions := range lost {
		for _, partition := range partitions {
			assignment := TopicPartition{topic, partition}

			c.mu.Lock()
			w := c.workers[assignment]
			delete(c.workers, assignment)
			c.mu.Unlock()

			close(w.stop)
			wg.Add(1)

			go func(a TopicPartition) {
				<-w.stopped

				wg.Done()
			}(assignment)
		}
	}
}

func (c *coordinator) run(ctx context.Context, cl *kgo.Client) {
	for {
		f := cl.PollFetches(ctx)
		if f.IsClientClosed() {
			return
		}

		f.EachError(func(topic string, part int32, err error) {
			c.logger.WithError(err).ErrorContext(ctx, "Problem occurred while fetching records")
		})

		f.EachPartition(func(p kgo.FetchTopicPartition) {
			// the library is giving zero records occasionally for a partition. Need to look into this to see
			// as it seems like a bug.
			if len(p.Records) == 0 {
				return
			}

			tp := TopicPartition{p.Topic, p.Partition}

			c.mu.RLock()
			w, ok := c.workers[tp]
			c.mu.RUnlock()
			if ok {
				c.sendRecordsToWorker(tp, cl, w, p)
			}
		})
	}
}

func (c *coordinator) sendRecordsToWorker(tp TopicPartition, client *kgo.Client, w *worker, f kgo.FetchTopicPartition) {
	w.records <- f.Records

	if _, pause := c.shouldPause(w); pause {
		// pause fetching more records for this partition until the worker finishes makes progress. NOTE
		// that the Franz client buffers records internally, so even after we pause this partition, we
		// may get a few batches delivered for this worker. This is the reason we needed a buffered channel with
		// a larger capacity; so we don't block delivering records to the worker
		client.PauseFetchPartitions(map[string][]int32{tp.Topic: {tp.Partition}})
	}
}

func (c *coordinator) shouldPause(w *worker) (int, bool) {
	l := len(w.records)
	return l, float32(l) > c.pauseThreshold
}

type worker struct {
	logger log.Logger

	// the Topic and Partition that this worker is handling records for
	assignment TopicPartition
	client     *kgo.Client
	listener   RecordListener

	// stop is used by the coordinator to ask a worker to stop what it is doing, and exit
	stop chan struct{}
	// stopped will be closed when this worker has stopped, i.e. terminated
	stopped chan struct{}
	records chan []*kgo.Record

	// this is the number of items in the records channel under which we will resume fetching data for this
	// [TopicPartition]
	resumeThreshold float32
}

func (w *worker) run(ctx context.Context) {
	defer func() {
		// we drain the records channel in the unlikely event that the fetch loop is blocked trying to put more
		// records into this channel
		drainChannel(w.records)
		close(w.stopped)
	}()

	for {
		select {
		case <-w.stop:
			return
		case recs := <-w.records:
			toCommit := w.listener.OnRecords(ctx, w.assignment, w.stop, recs)

			if err := w.client.CommitRecords(ctx, toCommit...); err != nil {
				w.logger.WithError(err).ErrorContext(ctx, "Problem committing records")
			}

			if w.shouldResume() {
				// NOTE that a worker that doesn't have much of a backlog will emit "extra" resumes which Franz
				// treats as no ops.
				w.client.ResumeFetchPartitions(map[string][]int32{w.assignment.Topic: {w.assignment.Partition}})
			}
		}
	}
}

func (w *worker) shouldResume() bool {
	return float32(len(w.records)) <= w.resumeThreshold
}

func drainChannel(ch <-chan []*kgo.Record) {
	for {
		select {
		case _, ok := <-ch:
			if !ok {
				return
			}
		default:
			return
		}
	}
}
```

```
	franzOpts = append(franzOpts,
		kgo.OnPartitionsAssigned(c.onPartitionsAssigned),
		kgo.OnPartitionsRevoked(c.onPartitionsLost),
		kgo.OnPartitionsLost(c.onPartitionsLost),
		// we are managing committing ourselves
		kgo.DisableAutoCommit(),
	)

**Comment by stampy88** [comment]
FYI, I can simplify the code above if it makes it easier to see what's going on.

**Comment by twmb** [comment]
Yes please minimize a bit. Also, I noticed that there actually is a race in v1.14's pausing, which is fixed in this branch: https://github.com/twmb/franz-go/pull/509
Can you try 509 and lmk if that fixes it? I plan to release it in a few hours.

**Comment by stampy88** [comment]
Yes, absolutely to both. I updated the code so that you can hopefully see any issues. I'll give 509 a try shortly.

**Comment by twmb** [comment]
* I'm confused why the onPartitionsAssigned is using a WaitGroup at all. The `Done` is being called right after the `Add(1)`.
* The library gives zero records when there is a partition error. Are you seeing no-error, no-record partitions?
* You can update (remove) the doc on pause about returning more records for a partition after pausing
* You should select on stopped when sending in sendRecordsToWorker
* drainChannel isn't the safest at the moment, and isn't necessary if you address the prior bullet point
* The c.mu lock can be removed if you use BlockRebalanceOnPoll (which will also make some of this easier to reason about)

**Comment by twmb** [comment]
* Running vs. about to be ran -- not too much of a difference -- some of the goroutines may be stopped to start the others :)
* Yes, `EachPartition` is always called on all partitions (with or without error)
* Good to hear about 509 -- that's everything I can think of as problematic with the release, so I'll go ahead and tag it and release. I hope you don't find something overnight 😬

**Comment by twmb** [comment]
Tagged v1.14.1, closing this -- please lmk if you're still experiencing issues

**Comment by stampy88** [comment]
Hi, @twmb, it happened again over the night. However, right before there are problems, I see "unable to join group session: UNKNOWN_MEMBER_ID: The coordinator is not aware of this member." errors returned by the `PollFetches`.

**Comment by stampy88** [comment]
No, this is without `BlockRebalanceOnPoll`; that is going in today. Let me look at logs to see if there was a `onPartitionsLost` that was going on at the same time that preceded the error, but didn't finish "quickly". Note that the `stop` channel is given to the application so that it can interrupt what it is going on the event a partition is lost: 
`toCommit := w.listener.OnRecords(ctx, w.assignment, w.stop, recs)` . The idea being that we don't want hold up a rebalance.

**Comment by stampy88** [comment]
Also, not all partitions failed after this. There are some still active. So it _seems_ like a race condition.

**Comment by twmb** [comment]
Can you turn on debug logs and paste a maybe slimmer sample of code?

I looped my own tests for a few hours before releasing to eliminate any seen race conditions, so I wonder if you're doing something that either I haven't tested / am not expecting, or if there's a small bug in the usage of the client (though the snippet above does not look buggy). The unknown_member_id is only returned from the broker when the client is not actually rebalancing, and given the code above, I expect your consumer is getting stuck in handlePartitionsLost because a worker isn't stopping.

Since this is a cooperative consumer, records can be returned from any non-lost partitions even during and after a rebalance -- until the member is finally booted. As well, if the client truly is stuck in handlePartitionsLost, then the following process can occur:
* Partitions actually lost, they are internally revoked, this calls OnPartitionsRevoked
* Your callback hangs
* Client continues consuming internally
* Another rebalance happens, but this cannot proceed until the prior one finishes
* Client cannot internally revoke lost partitions because the prior OnPartitionsRevoked has not returned
* You'll still see data for the lost partitions even though the client is no longer a member of the group
Can you log within your OnPartitionsAssigned / OnPartitionsLost to see that they're actually returning?

**Comment by stampy88** [comment]
So I do have a fair amount of logging around partitions being assigned/lost, so the `OnPartitionsLost` started right before that join error, and completely quickly (the numbers to the right are the partitions). An `OnPartitionsAssigned` occurred and that was fast as well:
```
"Jul 12, 2023 @ 03:31:52.492","Terminating workers for partitions...","3,4,10,15,16,22,27, 28"
"Jul 12, 2023 @ 03:31:52.655","Problem occurred while fetching records","-","-","unable to join group session: UNKNOWN_MEMBER_ID: The coordinator is not aware of this member."
"Jul 12, 2023 @ 03:31:52.655","Terminating workers for partitions...complete","3,4,10,15,16,22,27, 28"
"Jul 12, 2023 @ 03:31:55.832", "Creating workers for partitions assigned...", "25, 33, 41, 42, 3, 11, 12, 20"
"Jul 12, 2023 @ 03:31:55.832", "Creating workers for partitions assigned...complete", "25, 33, 41, 42, 3, 11, 12, 20"
```
I'll look to see if any pauses/resumes were going on around this as well.

**Comment by stampy88** [comment]
And yes, I'll post a working example of code.

**Comment by stampy88** [comment]
Partition 3 was paused prior to the partitions being lost. When the partition was re-assigned, could there be old baggage regarding 3 being paused that would cause it to subsequently never be fetched? Seems unlikely, but wanted to mention it.

**Comment by stampy88** [comment]
I don't see the paused topic/partitions being cleared for a partition we are losing: https://github.com/twmb/franz-go/blob/master/pkg/kgo/consumer_group.go#L538 Am I reading this correctly?

**Comment by twmb** [comment]
What do you mean by being cleared?

**Comment by stampy88** [comment]
I mean this `paused.delPartitions(topicPartitions)` is never done, it appears, when we lose a topic/partition. And this "old" pause data is causing the topic/partition never to be fetched next time it is assigned.

**Comment by twmb** [comment]
Yes, the paused state of a partition is independent of the partition being assigned: "Paused partitions persist until resumed" is a very hard persistence.

This guards against a person pausing a partition, losing it in a rebalance, having it reassigned in another rebalance, and for some reason automatically receiving data for that partition again (i.e. there is no opinionated decision to automatically unpause). If you want to unpause when losing the partition, you'll need to unpause the partition in OnPartitionsRevoked

**Comment by stampy88** [comment]
Hmmm, that is not how I expected it to work. I expected any client metadata around a topic/partition would be removed once the partition was lost. I'll make sure I resume it before OnPartitionsRevoked returns. You might want to consider updating the comment for the pause methods. Or I could do it as a first PR :)

**Comment by twmb** [comment]
Up to you what to clarify -- at most I can think that adding ", even if a rebalance occurs" could be more clarifying. Does unpausing in revoke fix things for you?

**Comment by stampy88** [comment]
Yeah, it would be something like "Paused partitions persist until resumed, even if a partition is lost during a rebalance. What this means is that if another rebalance occurs, and the pause partitioned is reassigned back to this consumer, it will still be paused". 

I am 99.9% sure that is the issue. I'll have a fix deployed some time today, so I'll let you know if it doesn't fix it. Thanks again for your help, @twmb. 

I do have a question around publishing in batch, but I can open a separate issue if you prefer.

**Comment by twmb** [comment]
Sure, separate issue. I'll close this one.

**Comment by vtolstov** [comment]
> Hi, @twmb, it happened again over the night. However, right before there are problems, I see "unable to join group session: UNKNOWN_MEMBER_ID: The coordinator is not aware of this member." errors returned by the `PollFetches`.

I have such issue too. But i'm using old v1.12.1 package. I'm unable to debug issue now, because issue appeared only after 5-6 hours of intensive consuming from kafka. And issue appears mostly then i have 1-2s of processing time for each message (consumer calls firebase server via rest that sometimes slowdowns...)

**Comment by twmb** [comment]
Please update to latest and let me know if it continues to happen
